### DISCUSSION TITLE
Oauth "The redirect URI included is not valid"

### DISCUSSION THREAD

**Discussion by Slonskiy** [okd-project/okd#1322] [Q&A]
**Describe the bug**
I'm trying to use a GitLab identity provider as described in official documentation:
https://docs.okd.io/latest/authentication/identity_providers/configuring-gitlab-identity-provider.html
I choose gitlab at console form, redirecting to a gitlab, successful authorize there, then at gitlab form I get "The redirect URI included is not valid."

Looks like all resources created normally (I was created them via web-console):
oc describe OAuth cluster                                   
```
Name:         cluster
Namespace:    
Labels:       <none>
Annotations:  include.release.openshift.io/ibm-cloud-managed: true
              include.release.openshift.io/self-managed-high-availability: true
              include.release.openshift.io/single-node-developer: true
              release.openshift.io/create-only: true
API Version:  config.openshift.io/v1
Kind:         OAuth
Metadata:
  Creation Timestamp:  2022-08-30T14:38:20Z
  Generation:          3
  Managed Fields:
    API Version:  config.openshift.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:include.release.openshift.io/ibm-cloud-managed:
          f:include.release.openshift.io/self-managed-high-availability:
          f:include.release.openshift.io/single-node-developer:
          f:release.openshift.io/create-only:
        f:ownerReferences:
          .:
          k:{"uid":"5c05a64d-ee01-46f1-bc06-3323de124f83"}:
      f:spec:
    Manager:      cluster-version-operator
    Operation:    Update
    Time:         2022-08-30T14:38:20Z
    API Version:  config.openshift.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        f:identityProviders:
    Manager:    Mozilla
    Operation:  Update
    Time:       2022-08-30T15:12:00Z
  Owner References:
    API Version:     config.openshift.io/v1
    Kind:            ClusterVersion
    Name:            version
    UID:             5c05a64d-ee01-46f1-bc06-3323de124f83
  Resource Version:  116870
  UID:               0262aea8-c82a-40f2-89d9-5fab06619291
Spec:
  Identity Providers:
    Gitlab:
      Ca:
        Name:     ca-cf-config-map
      Client ID:  66b26bca84f5e8a5d1bc4e06824cc78940f355dd6462a41734cf9c4299286828
      Client Secret:
        Name:        gitlab-client-secret-9sv46
      URL:           https://gitlab.our.domain/
    Mapping Method:  claim
    Name:            gitlab
    Type:            GitLab
Events:              <none>
```

Secret and config map with CA is present:
**oc get secret gitlab-client-secret-9sv46 -n openshift-config**
NAME                         TYPE     DATA   AGE
gitlab-client-secret-9sv46   Opaque   1      21h
**oc get cm ca-cf-config-map -n openshift-config**              
NAME               DATA   AGE
ca-cf-config-map   1      19h

URI is made by template from documentation, e. g.
https://oauth-openshift.apps.cluster_name.cluster_domain.com/oauth2callback/GitLab 

And I'm sure their contents are relative (Client ID, clientSecret and root CA). Other applications using gitlab oauth is authenticating well. 

I'm actually not sure that it's a bug... But have no idea how debug this situation. Is it possible than callback URI had changed at any version of OKD?

P. S. If it's matters - callback URI is available from internet, I see it when try to open it by browser or do curl request.

**Version**
4.11.0-0.okd-2022-08-20-022919
UPI, bare metal

**How reproducible**
Setup application in gitlab, create gitlab identity provider with relevant properties and try to log on via gitlab.

**Comment by vrutkovs** [comment]
This not a bug, this is a support request.

**Comment by vrutkovs** [comment]
> Name:            gitlab

but an uppercase name is used:

> URI is made by template from documentation, e. g.
https://oauth-openshift.apps.cluster_name.cluster_domain.com/oauth2callback/GitLab

Check if using a lowecase name:
https://oauth-openshift.apps.cluster_name.cluster_domain.com/oauth2callback/gitlab

works?

**Reply by Slonskiy** [reply]
Got changes, but auth isn't successful:  "An authentication error occurred." on OKD page (after redirecting from gitlab).

I tried copy/paste url from browser and request it with curl (but not sure that's correct, token can expired at request moment):
**curl --cacert Cloudflare_CA.pem --cookie "csrf=YRnYTPB3vY2ghPwGPhz9tDxIB8olyqIBxNx3q1p_3FU" 'https://oauth-openshift.apps.cluster_name.cluster_domain.com/Foauth2callback/gitlab/response_type=code&scope=openid&state=Y3NyZj1QXzdlRENyUmh6cTZJd0dVX3V2UWg3dXJzZjg4WWFjUlktQ3pVbUxRVlJZJnRoZW49JTJGb2F1dGglMkZhdXRob3JpemUlM0ZjbGllbnRfaWQlM0Rjb25zb2xlJTI2aWRwJTNEZ2l0bGFiJTI2cmVkaXJlY3RfdXJpJTNEaHR0cHMlMjUzQSUyNTJGJTI1MkZjb25zb2xlLW9wZW5zaGlmdC1jb25zb2xlLmFwcHMub2NwNC5nYXphbWJhLmNvbSUyNTJGYXV0aCUyNTJGY2FsbGJhY2slMjZyZXNwb25zZV90eXBlJTNEY29kZSUyNnNjb3BlJTNEdXNlciUyNTNBZnVsbCUyNnN0YXRlJTNEYmRiNjNkMzM%3D'**
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "forbidden: User \"system:anonymous\" cannot get path \"/Foauth2callback/gitlab/response_type=code\u0026amp;scope=openid\u0026amp;state=Y3NyZj1QXzdlRENyUmh6cTZJd0dVX3V2UWg3dXJzZjg4WWFjUlktQ3pVbUxRVlJZJnRoZW49JTJGb2F1dGglMkZhdXRob3JpemUlM0ZjbGllbnRfaWQlM0Rjb25zb2xlJTI2aWRwJTNEZ2l0bGFiJTI2cmVkaXJlY3RfdXJpJTNEaHR0cHMlMjUzQSUyNTJGJTI1MkZjb25zb2xlLW9wZW5zaGlmdC1jb25zb2xlLmFwcHMub2NwNC5nYXphbWJhLmNvbSUyNTJGYXV0aCUyNTJGY2FsbGJhY2slMjZyZXNwb25zZV90eXBlJTNEY29kZSUyNnNjb3BlJTNEdXNlciUyNTNBZnVsbCUyNnN0YXRlJTNEYmRiNjNkMzM=\"",
  "reason": "Forbidden",
  "details": {
    
  },
  "code": 403
} 

**UPD**: 
If I do curl request faster (probably token isn't expired), I got the same message at console, like in browser:
**curl --cacert Cloudflare_CA.pem --cookie "csrf=YRnYTPB3vY2ghPwGPhz9tDxIB8olyqIBxNx3q1p_3FU" 'https://oauth-openshift.apps.cluster_name.cluster_domain.com/oauth2callback/gitlab?code=f054b8888917309427a0c4cb6fcdbf6a57c1448564813a2c0a49ba70e1fe4ec4&state=Y3NyZj1ZUm5ZVFBCM3ZZMmdoUHdHUGh6OXREeElCOG9seXFJQnhOeDNxMXBfM0ZVJnRoZW49JTJGb2F1dGglMkZhdXRob3JpemUlM0ZjbGllbnRfaWQlM0Rjb25zb2xlJTI2aWRwJTNEZ2l0bGFiJTI2cmVkaXJlY3RfdXJpJTNEaHR0cHMlMjUzQSUyNTJGJTI1MkZjb25zb2xlLW9wZW5zaGlmdC1jb25zb2xlLmFwcHMub2NwNC5nYXphbWJhLmNvbSUyNTJGYXV0aCUyNTJGY2FsbGJhY2slMjZyZXNwb25zZV90eXBlJTNEY29kZSUyNnNjb3BlJTNEdXNlciUyNTNBZnVsbCUyNnN0YXRlJTNEN2Y5N2ZkMDc%3D'** 
An error occurred

And no more information.

**UPD**
I think I understood now. At first it was wrong callback URI, now have some problems with certificate. 
At logs one of oauth pods:
```
I0901 06:35:35.116973 1 handler.go:174] Error getting access token from an external OIDC provider (https://gitlab.domain.de/oauth/token?client_id=66b26bca84f5e8a5d1bc4e06824cc78940f355dd6462a41734cf9c4299286828&client_secret=1aba5c8e394366661df450df3fb2b08b7edcf904846ba8319e12c5f34c32b827&code=f054b8888917309427a0c4cb6fcdbf6a57c1448564813a2c0a49ba70e1fe4ec4&grant_type=authorization_code&redirect_uri=https%3A%2F%2Foauth-openshift.apps.cluster_name.cluster_domain.com%2Foauth2callback%2Fgitlab): Post "https://gitlab.domain.de/oauth/token": x509: certificate signed by unknown authority
E0901 06:35:35.117057 1 errorpage.go:28] AuthenticationError: Post "https://gitlab.domain.de/oauth/token": x509: certificate signed by unknown authority
```

Thanks for help! 
And maybe it's the reason to add example of callback URI to the manual? Because I used spec.identityProviders.type as a value to 'idp-provider-name' . In my opinion, manual may be more oblivious with that.
